libr/bin/format/elf/elf.c

```diff
@@ -293,18 +295,67 @@ static int init_dynamic_section (struct Elf_(r_bin_elf_obj_t) *bin) {
 		return false;
 	if (bin->phdr[i].p_offset + sizeof(Elf_(Dyn)) > bin->size)
 		return false;
-	tmp = dyn = (Elf_(Dyn)*)((ut8 *)bin->b->buf + bin->phdr[i].p_offset);
-	for (entries = 0; (ut8*)dyn < ((ut8*)tmp + dyn_size); dyn++) {
-	    	entries++;
-		if (dyn->d_tag == DT_NULL) {
+
+	tmp = bin->phdr[i].p_offset;
+	if (tmp < bin->b->base || tmp > (bin->b->base + bin->b->length)) {
+		if (!(iob && iob->io)) {
+			return false;
+		}
+		if (tmp + bin->b->length > bin->size) {
+			if (iob->read_at (iob->io, bin->size - bin->b->length, bin->b->buf, bin->b->length) < bin->b->length) {
+				return false;
+			}
+			bin->b->base = bin->size - bin->b->length;
+			bin->b->cur = bin->size - tmp;
+			tmp -= bin->b->base;
+		} else {
+			if (iob->read_at (iob->io, tmp, bin->b->buf, bin->b->length) < bin->b->length) {
+				return false;
+			}
+			bin->b->base = tmp;
+			bin->b->cur = 0;
+			tmp = 0;
+		}
+	}
+	for (entries = 0; (entries * sizeof (Elf_(Dyn))) < dyn_size; ) {
+		entries++;
+		if (((Elf_(Dyn)*)((ut8*)bin->b->buf + tmp))->d_tag == DT_NULL) {
 			break;
 		}
-		if ((ut8*)(dyn+2) > ((ut8*)bin->b->buf + bin->size))
-		    	return false;
+		if ((bin->b->base + tmp + sizeof(Elf_(Dyn))) > bin->size) {
```

Comment:
this would lead to recopying of buffer. I don't know which one is better: copying buffer again and again or using the existing buffer and do seek+read in that.